gradio/route_utils.py

```diff
@@ -583,3 +584,40 @@ def starts_with_protocol(string: str) -> bool:
     """
     pattern = r"^[a-zA-Z][a-zA-Z0-9+\-.]*://"
     return re.match(pattern, string) is not None
+
+
+class CustomCORSMiddleware(BaseHTTPMiddleware):
+    async def dispatch(self, request: fastapi.Request, call_next):
+        # Extract host and origin from the request headers
+        host_header: str = request.headers.get("host", "")
+        origin_header: str = (
+            request.headers.get("origin", "")
+            .replace("http://", "")
+            .replace("https://", "")
+        )
```

Comment:
You'll probably want to use `starlette.datastructures.URL` here.

Reply:
Yeah fair enough

Reply:
Fixed